### PR TITLE
traverse: add topo

### PR DESCRIPTION
**Pull request by haampie**
@scheibelp since we talked about this. It's useful in the installer, as well as for removal (currently it screams about the db being in an inconsistent state), and as you suggested: environment modifications (although I'm still not convinced).

Basically gives you 

```python
spack.traverse.traverse_edges([s1, s2], order="topo", direction="parents/children", key=...)
```

and `spack.traverse.traverse_nodes`.

### CODE REVIEW

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [inline lib/spack/spack/test/traverse.py]: This should be represented as a more-traditional ascii graph (i.e. each node appears once). I assume that was not done here because it is painful. I wonder if `spack.graph.AsciiGraph` could be generalized for this purpose; in this particular case I don't think you would need to modify the function at all though, because you create the DAG using `Spec` objects (which is what `graph` wants).

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [inline lib/spack/spack/test/traverse.py]: (minor) I'm assuming you're doing this to turn a generator into a re-traverse-able object: it's more idiomatic to do that like `specs = list(ordered_specs)` but IMO clearer to fold that into the prior line (to avoid more variable names).

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [inline lib/spack/spack/test/traverse.py]: Why set `order`? In general, if a parameter's value shouldn't matter, it shouldn't be set.

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [inline lib/spack/spack/test/traverse.py]: This looks similar to https://en.wikipedia.org/wiki/Topological_sorting, I'll note that IMO this doesn't sufficiently cover the types of cases we want to get right with topological ordering; in fact I don't think it contains any subgraphs that BFS wouldn't traverse in the same way (in particular if you started from a "pseudo-root" that attached to all the disconnected roots. An example a subgraph that ought to be in this test:

```
0
| \
1 |
| /
2 
```

i.e. there should be "shortcuts" to child nodes. Ideally this type of structure would be common (and repeated a few times at different scales) in the graph used for testing.

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py:113]: It is unclear to me how this is achieving a topological order. I think the docstring should clarify that (it's probably worth talking about the general strategy here first though). If I had to guess, I'd say it's based on the depth-first-derived strategy discussed in https://en.wikipedia.org/wiki/Topological_sorting; it's unclear to me how composing visitors achieves that though.

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: I'm not sure what this comment applies to: the function starts by iterating over nodes. Also I think this should explain what the "standard implementation" is (or for example it could be worded like "instead of doing X, do Y because...").

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py:423]: (question) what was the original purpose of being able to specify the difference between a pre-order and post-order DFS traversal? (I assume you don't have the original context, I want to resolve this though before merging the PR and I'll need to track other folks down).

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: If I understand correctly, these are `EdgeAndDepth` instances. This isn't strictly an idiom established by this PR, but calling these `node`s is confusing to me, is there any way that could be changed? (if so, perhaps that ought to happen in a separate PR)

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: This should have a docstring that explains `direction="children"` means that parents are ordered before children. I'm inclined to think the parameter should be renamed though: to something like `prioritize="parents"`.

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: For now, I think it makes sense to prohibit `cover="edges"` for `order="topo"`: I don't know of an interpretation of topological ordering which applies for edges.

**Review by scheibelp — CHANGES_REQUESTED** [review]
At a high-level:

* The implementation needs documentation (I'm not clear how composing visitors achieves a topological ordering)
* The test graph needs to be more complex
* Given that pre/post order relate to DFS, I want to check the expectations of callers of `traverse` in the code: did they actually mean "topo" when they were asking for pre/post? This isn't something you need to investigate yourself (although that would be helpful)

File: lib/spack/spack/test/traverse.py
Comment (haampie) [reply lib/spack/spack/test/traverse.py]: It matters cause it shouldn't be `topo`

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py]: In textbook dfs you pass a vertex not a list of edges

File: lib/spack/spack/test/traverse.py
Comment (haampie) [reply lib/spack/spack/test/traverse.py]: Nice, it's a bit unfortunate that it works on a single Spec instead of actually accepting all nodes or at least multiple roots that induce the full dag.

File: lib/spack/spack/test/traverse.py
Comment (haampie) [reply lib/spack/spack/test/traverse.py]: I see what you mean, but the example above would not be a failing test case for bfs either ;p

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py:113]: Hm, I thought it was reasonably clear (traverse all possible paths from A -> B before pushing A), but I can add a comment. Edit: added the reference

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py]: I don't agree with renaming it. Our vertices are specs, and the edges are directed from parent to child or dependent to dependency. Then this topo traversal code follows the definition of topo order. `direction="parents"` is there to flip the edges, and it's consistent with traverse_edges(...) which we've had forever.

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py:423]: Typically for undoing actions, for example pushing (in pre) and popping (in post) items on a stack.

Edit: naively I thought we make environment variable changes in post order and then revert those changes in pre order, but `reverse(pre) != post`, so it would not always be correct.

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py]: It's ~not hard~ may not be useful to define? For a set of edges `{u -> v}`, define the binary operator `(s -> t) < (u -> v)` as `t  < u` where `<` induces the topo order on the set of vertices.

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py]: It's a remnant of refactoring I think. Going from a function that took a vertex and then looped over its neighboring vertices, to a function that took a list of neighbors :thinking:

Comment (haampie) [reply lib/spack/spack/traverse.py]: Feel free to just submit a pr with these small things next time, then I can already merge it the next morning and rebase my work on it.

File: lib/spack/spack/test/traverse.py
Comment (haampie) [inline lib/spack/spack/test/traverse.py]:
```suggestion
    # edges are ordered by target node name. Roots are {A, E} in forward order and
```

**Comment by haampie** [comment]
@scheibelp my main Q is whether it fits the `traverse_edges` interface enough. I don't think we need `cover=edges` necessarily, but one way to think about it is: we yield all edges of the DAG, in such an order that all in-edges of a vertex are yielded before all out-edges.

Edit: I just pushed that.

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py]: In the latest commits `cover=edges` + `order=topo` generates all edges of the DAG ordered s.t. all in-edges per vertex precede its out-edges.

File: lib/spack/spack/traverse.py
Comment (scheibelp) [reply lib/spack/spack/traverse.py]: Not renaming is fine. In that case, the docstring should explain the meaning of `direction="children"`

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: Typo: "ever" -> "every"

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: Should "we only append a vertex A before" be "we only append a vertex A *after*"?

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [reply lib/spack/spack/test/traverse.py]: I think you mean you don't want to rely on "topo" to retrieve the parents while testing the ordering of topo (`topo` should only affect the ordering: any traversal of a node - `pre`, `post`, `topo`, or `bfs` - should produce the same set of nodes).

If I got that right, is there a test that exists which confirms that all the ordering parameters produce the same set of parents? I think that would be useful (and in that case this test wouldn't have to specify the ordering).

File: lib/spack/spack/traverse.py
Comment (scheibelp) [reply lib/spack/spack/traverse.py]: I'm generally reluctant to implement something we have no known use case for, but does not implementing this cause some other issue (i.e. a failure since the code using this makes assumptions about what is available in the interface)?

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: > The neat thing about this algorithm is that it can operate on an immutable DAG, it doesn't require marking or popping edges, and it works in linear time (in vertices and edges).

This particular statement is hard to understand in this context: it is useful for comparing implementations of topological sort, but we only have one here. Also AFAIK the `visited` set effectively functions as a form of marking, so I think whether this statement applies is also debatable. Either way, I think the docstring is clearer without it.

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: And `post()`

IMO this nested composition of visitors in tandem with specialized user methods is more confusing than an "all-in-one" implementation of topological sorting: having to combine the behavior of this method with the implementation of `TopoVisitor` is confusing to me (on that note, I think `TopoVisitor` should specify that it produces a topological ordering when combined with this function, so that future readers know to examine both when trying to determine the semantics).

File: lib/spack/spack/traverse.py
Comment (scheibelp) [reply lib/spack/spack/traverse.py]: I think the following description would be clearer, but I'm happy to iterate on what's here:

Perform a depth-first search; append a node only after visiting all of its children. This produces a list where the children of a node are always guaranteed to come before a node. This list can be reversed to produce a topological ordering.

Again, this depends on being used in tandem with `traverse_depth_first_with_visitor`, and to that end I think it would be clearer if this class were defined inside of the method body of `traverse_edges_topo`.

**Review by scheibelp — COMMENTED** [review]
Additional comments

File: lib/spack/spack/test/traverse.py
Comment (haampie) [reply lib/spack/spack/test/traverse.py]: Actually another reason why I did this is because `root=True/False` was not implemented, but it is now.

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py]: I don't agree with the latter, it's not an exposed class in the module, but for advanced use cases you might want to override parts of it, like `neighbors`. So, not public API, but it can be useful nonetheless.

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py]: Hm yeah, let me just drop it. It was in contrast to the Kahn algorithm where it's common delete edges, but even that can run in linear time if you keep a dict of `node id` -> `# in-edges`.

Btw, notice that if we would do Kahn instead of this DFS/backtracking algorithm, we would still need to do one pass with BFS/DFS to actually get the `# in-edges` count of the sub-DAG we're traversing over. So I think this DFS algorithm is typically more efficient.

File: lib/spack/spack/traverse.py
Comment (haampie) [reply lib/spack/spack/traverse.py]: I get your point, but I disagree, because there are instances where the `traverse_nodes` interface is too limiting, for example because `deptype` is not constant but depends on depth (drop build edges at depth > 0) or the properties of the node. Also, the composability aspect is nice, and in principle it's possible to unit test the behavior of a visitor (for instance one that has a stateful `neighbors()` implementation).

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: Note that by itself, the constraint "in-edges before out-edges" does not produce a topological order:

```
0
|\
1 |
|/
3
```

`[0, 3], [0, 1], [1, 3]` satisfies that but is not a valid topological ordering. The comment should probably first introduce the traditional concept of topological ordering and then extend it to edges.

This is partly my concern with generalizing the notion of topological ordering to edges: suddenly the documentation speaks in terms that are unfamiliar and need to be reasoned about, vs. concepts which are immediately accessible (e.g. via a Wikipedia article) - reorganizing the docstring would help with this though.

Comment (haampie) [reply lib/spack/spack/traverse.py]: I agree. I'm not sure how useful it is to allow cover=edges at this point. You do have the property that `map(lambda s: s.parent, traverse_edges(order="topo))` is topo ordered, but yeah, that does not include all reachable vertices.

Comment (scheibelp) [reply lib/spack/spack/traverse.py]: Sorry to be clear I trusted that this produces a topological order (or that one can be extracted from the list of edges), my concern is that it will be hard for future maintainers to understand it; the invariant stated in the docstring is insufficient to guarantee a topological ordering (the algorithm itself does more than that though), which I think is an example of how it's challenging to describe the semantics.

Are you inclined to undo the support for `cover=edges`, or would you prefer to improve the docstrings? Either way is fine with me.

Comment (haampie) [reply lib/spack/spack/traverse.py]: I just updated the docstrings and kept `cover=edges`

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [inline lib/spack/spack/test/traverse.py]: Add comment explaining what this is testing (i.e. this is just testing the in-edges/out-edges ordering on a per-node basis, and the other test is responsible for verifying the traditional definition of a topological ordering).

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [inline lib/spack/spack/test/traverse.py:339]: I don't think DFS specifies whether C or B would be visited first from A, so it is more accurate to say that this is a valid DFS visitation order but not a valid topological order. I think this is important to mention because it clarifies that this is hard to test: I'm not sure if we can make a graph where DFS is *guaranteed* to produce an order that is not topological, so in that sense this test is a heuristic.

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [inline lib/spack/spack/test/traverse.py:385]: I think it's worth mentioning that if some root specs are children of others, the resulting traversal is effectively the same as if they had not been included [1]. This might be better to put in the docstring of `traverse_nodes` than here.

1. Except for depth: Reading the docstring of `traverse_nodes`, it mentions that the roots are considered depth 0: what does that mean in this example where D is a child of A (i.e., what is the depth of B)? That conceptual difficulty is not strictly related to any of the changes here, but I bring it up because I had to spend some time thinking about what it means for a topological ordering when some roots are children of others; I think it works fine here but introduces potential extra complexity for future maintainers.

File: lib/spack/spack/traverse.py
Comment (scheibelp) [reply lib/spack/spack/traverse.py:423]: Thanks for checking into that! Any issues with pre/post elsewhere can be resolved separately and I consider this chain resolved for the purposes of this PR.

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: In this case I assume that all edges (both in and out) involving an excluded vertex would be discarded. I think the issue is not that we cannot define a reasonable interpretation, but that we have no use cases for topological traversals of edges.

Therefore, I think this comment should be rephrased:

```
# TODO: There is no known need for topological ordering of traversals (edge or node)
# with an initialized "visited" set. Revisit if needed.

File: lib/spack/spack/traverse.py
Comment (scheibelp) [inline lib/spack/spack/traverse.py]: (phrasing) 

I think

> Returns a list of edges where ordered in topo order...

should be 

> Returns a list of edges in topological order...

**Review by scheibelp — CHANGES_REQUESTED** [review]
The remaining requests are all for documentation (although in one case I suggest that we may want to constrain the set of roots allowed for `traverse_nodes`).

File: lib/spack/spack/test/traverse.py
Comment (haampie) [reply lib/spack/spack/test/traverse.py:339]: It is actually an invariant of `traverse_nodes(order="pre/post")` to order edges by target package name. The comment hints on that. I can reword it to make that more clear.

File: lib/spack/spack/test/traverse.py
Comment (haampie) [reply lib/spack/spack/test/traverse.py:385]: They shouldn't be called roots if they have in-degree != 0 :man_shrugging: that's also how I'm phrasing it in the comment, but I can expand the comment. In `traverse_nodes` the positional argument is called `specs` not `roots`. It's just a bit unfortunate we have a kwarg named `root=True`, and it's always used to mean "skip the traversal starting point"; maybe `skip_start=True/False` was better. I guess `specs` is more like an ordered list of children of an artificial root (and that root induces the sub-dag of interest), so even then the input specs are not roots :p.

Notice though that in most use cases before we did something like this:

```python
visited = set()
for s in specs:
  for t in s.traverse(root=False, visited=visited):
    # will give you specs[1] if there's an edge specs[0] -> specs[1]
```

which is now equivalent to

```python
for t in traverse(specs, root=False):
    # will give you specs[1] if there's an edge specs[0] -> specs[1]
```

So, the "potential extra complexity" was already an issue before.

That complexity is gone when using `order=breadth`, since then you know for a fact that non-root input specs are discovered at depth 0.

Comment (haampie) [reply lib/spack/spack/test/traverse.py:385]: So, I think I can clarify that elsewhere, but maybe in a separate PR. For now I've just updated the relevant comment in the test.

Comment (haampie) [reply lib/spack/spack/test/traverse.py:385]: Aargh, I realize I *did* use the word `roots` in the test. Will fix that too.

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [reply lib/spack/spack/test/traverse.py:339]: That is good to know: I think this test is much better than I originally thought. However, I don't see any mention of this invariant in the docstring for `traverse_nodes` or in `BaseVisitor.neighbors` (which appears to be the source of this invariant).

Comment (haampie) [reply lib/spack/spack/test/traverse.py:339]: See also https://github.com/spack/spack/pull/33406#discussion_r1007822909 and the linked PR https://github.com/spack/spack/pull/28673#discussion_r823132555. If I remember correctly, the idea was that this was an invariant thanks to the Spec objects themselves (EdgeMap is always sorted), but in practice even the old traversal code uses an explicit sort, that I didn't remove when refactoring to visitors.

Comment (scheibelp) [reply lib/spack/spack/test/traverse.py:339]: I agree that depending on the sorting properties of the edges is sufficient. I think it may be an issue though that we don't express that we use this property to guarantee a fixed, repeatable ordering of edges/nodes. This issue in documentation may predate this PR. It doesn't need to be fixed here.

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [reply lib/spack/spack/test/traverse.py:385]: > So, the "potential extra complexity" was already an issue before.

I wouldn't consider that the same kind of complexity: I was worried about semantics within a single traversal. Multiple traversals might add other complications.

> That complexity is gone when using order=breadth, since then you know for a fact that non-root input specs are discovered at depth 0.

To clarify, my specific example was not about a non-root input spec, but rather a spec which is a child of two input specs, appearing at different depths WRT each. Nevertheless, I think I see how BFS produces a well-defined (but meaningless) value for the depth of B when `specs=[A, D]`. However, since we allow `specs=[A, D]` for topo,  it should also make sense for that, or it should be forbidden (the latter being my preference). 

Overall, I'm noticing a trend where allowing `specs` to contain nodes that are children of other nodes in `specs` is making it harder to reason about the semantics; I think forbidding this is appealing because I don't know of any use case that requires it (and supporting it seems to add complexity).

> They shouldn't be called roots if they have in-degree != 0

I'm fine with not calling them roots (although don't really see why it's an issue to call them that) although preferably something more descriptive than `specs`.

> In traverse_nodes the positional argument is called specs not roots.
> Aargh, I realize I did use the word roots in the test. Will fix that too.

And in the docstring for traverse_nodes itself:

`specs (list): List of root specs (considered to be depth 0)`

they may not be roots in a strict sense, but the term "root" captures what we intend to do with it in a single (and short) word.

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [reply lib/spack/spack/test/traverse.py:339]: (question) would it also be required to sort `traverse.root_specs` to get a fixed ordering?

File: lib/spack/spack/test/traverse.py
Comment (haampie) [reply lib/spack/spack/test/traverse.py:385]: I don't really get the point you're making. In environments you may want to pass a list of env roots to topo traverse, in uninstall you wanna pass a list of specs matching a db query, etc. You don't know if one is a descendant of the other, that requires traversal. Forbidding it is only annoying. What we care about is topo traversal of the subdag induced by the list of input specs.

File: lib/spack/spack/test/traverse.py
Comment (haampie) [reply lib/spack/spack/test/traverse.py:339]: I don't want to sort them, since there are cases where there's a natural order such as with environment roots

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [reply lib/spack/spack/test/traverse.py:385]: > in uninstall you wanna pass a list of specs matching a db query

That is an excellent point (i.e. some specs might be transitive children of others). It's kind of a headache thinking about the meaning of multi-root traversals for the different traversal methods, but given this example it is certainly worthwhile for topological ordering.

File: lib/spack/spack/test/traverse.py
Comment (scheibelp) [reply lib/spack/spack/test/traverse.py:339]: This isn't strictly problematic since a user could sort this themselves before passing it in, but complicates the interface since its documentation should mention that the input specs should be pre-sorted to get a fixed ordering. I didn't anticipate a need for users to control traversal orders by sorting the specs themselves (is there a reason for example that users would want the environment spec listing order to influence the traversal order?).

**Review by scheibelp — APPROVED** [review]

**Pull request merged**
